Redesign plane resource types

rynowak opened the pull request:
# Description

This change splits the plane resource type into multiple types for maintainability as reviewed in radius-project/design-notes#42. See the design document for the motivation and dicussion of the specific design details. In summary, previously we had one resource type for all planes, and now we have a few separate ones.

This is a technically a breaking change but the impact on users is very low because it's mainly UCP itself that interacts with the planes.

The majority of the code changes are due to:

- Simplifying the resource group client API
- Updating and adding examples



## Type of change

- This pull request adds or changes features of Radius and has an approved issue (issue link required).


Fixes: #issue_number

rynowak commented:
This is a large PR due to the number of changes required when we introduce new APIs. This change is introducing 3 new APIs! I'd recommend starting with the [typespec](https://github.com/radius-project/radius/pull/7499/files#diff-d587dda48bb68ff15ac318818ac05035191faafb16dbf15c3301bca0efa8f5a6) and then looking at the [routes](https://github.com/radius-project/radius/pull/7499/files#diff-c93ff36d2c9bd07a997732d8ad4c5047f6dcf81cb265b0d36d593e4f2de4ae4d) for the the new APIs.

rynowak commented on docs/contributing/contributing-code/contributing-code-control-plane/configSettings.md:
```diff
@@ -142,7 +142,7 @@ ucp:
 | Key | Description | Example |
 |-----|-------------|---------|
 | id | The ID of the UCP plane | `/planes/radius/local` | 
-| type | The type of UCP plane | `System.Planes/radius` |
```
FYI I did a find and replace. There are no remaining uses of `System.Planes`.

rynowak commented on pkg/cli/clients/clients.go:
```diff
@@ -164,10 +164,10 @@ type ApplicationsManagementClient interface {
 	ListEnvironmentsAll(ctx context.Context) ([]corerp.EnvironmentResource, error)
 	GetEnvDetails(ctx context.Context, envName string) (corerp.EnvironmentResource, error)
 	DeleteEnv(ctx context.Context, envName string) (bool, error)
-	CreateUCPGroup(ctx context.Context, planeType string, planeName string, resourceGroupName string, resourceGroup ucp_v20231001preview.ResourceGroupResource) error
```
This is one of the more viral changes. Previous the resource group API documented that you could create a resource group in any plane. This is clearly false, it's just the `radius` plane. I fixed that mistake in the typespec which resulted in a change everywhere the resource group client is used.

rynowak commented on pkg/ucp/backend/controller/resourcegroups/trackedresourceprocess_test.go:
```diff
@@ -147,28 +144,6 @@ func Test_Run(t *testing.T) {
 		require.NoError(t, err)
 	})
-
-	t.Run("Failure (validate downstream: invalid downstream)", func(t *testing.T) {
```
This is an error case that is no longer possible. We previously needed to guard against a misconfigured `radius` plane everywhere that UCP interacts it.

rynowak commented on pkg/ucp/frontend/api/routes.go:
```diff
@@ -131,64 +125,17 @@ func Register(ctx context.Context, router chi.Router, planeModules []modules.Ini
 			OperationType:     &v1.OperationType{Type: OperationTypePlanes, Method: v1.OperationList},
 			ControllerFactory: planes_ctrl.NewListPlanes,
 		},
-		{
```
Now that each plane has it's own converters, the route registration for the plane APIS needed to be plane-specific. This is now handled by the individual modules for each plane type.

rynowak commented on pkg/ucp/frontend/controller/planes/listplanes.go:
```diff
@@ -31,18 +31,18 @@ import (
 
 var _ armrpc_controller.Controller = (*ListPlanes)(nil)
 
-// ListPlanes is the controller implementation to get the list of UCP planes.
```
This is a special case that we can't use the shared controller for. This is the `/planes` endpoint (list all planes regardless of type).

rynowak commented on pkg/ucp/integrationtests/planes/testdata/awsplane_updated_v20231001preview_requestbody.json:
```diff
@@ -1,7 +1,8 @@
 {
-    "location": "global",
```
Lots of these examples had mistakes, or were just missing things that should be there. I did a pass and make sure we were consistent in how we're testing.

Most of UCP testing is done in integration tests since it's a single "unit".

rynowak commented:
@youngbupark @ytimocin can you take a look when you have a chance? I need to figure out my one failing test, but that is likely a swagger update I missed.

youngbupark reviewed (commented):
Since this is a huge update, it is not easy to review the entire code thoroughly. Overall looks good to me. I will have some time to review this PR in detail tomorrow. meanwhile, please fix the test :)

ytimocin commented on pkg/ucp/datamodel/converter/azureplane_converter.go:
```diff
@@ -0,0 +1,59 @@
+			return nil, err
+		}
+		return versioned, nil
+
```
nit: extra line can be removed

ytimocin commented on pkg/ucp/datamodel/radiusplane.go:
```diff
@@ -22,33 +22,32 @@ import (
 	return p.Type
 }
 
 // LookupResourceProvider checks if the input provider is in the list of configured providers.
```
**Checks and returns**

rynowak replied:
Can you elaborate?

ytimocin commented on pkg/ucp/frontend/api/routes.go:
```diff
@@ -37,8 +34,8 @@ import (
-	planeCollectionPath       = "/planes"
-	planeCollectionByTypePath = "/planes/{planeType}"
+	planeCollectionPath      = "/planes"
+	planesTypeCollectionPath = "/planes/{planeType}"
```
Any reason why this is plural and the other is singular?

rynowak replied:
I think just a typo, I'll change it 👍

ytimocin commented on pkg/ucp/frontend/api/routes_test.go:
```diff
@@ -26,6 +26,11 @@ import (
 	"github.com/radius-project/radius/pkg/armrpc/rpctest"
 	"github.com/radius-project/radius/pkg/ucp/dataprovider"
 	"github.com/radius-project/radius/pkg/ucp/frontend/modules"
+<<<<<<< HEAD
```
Maybe a bad merge?

rynowak replied:
Yes :) Working on fixing it.

ytimocin commented on pkg/ucp/frontend/api/server.go:
```diff
@@ -190,51 +192,88 @@ func (s *Service) Initialize(ctx context.Context) (*http.Server, error) {
+				RequestConverter:  converter.RadiusPlaneDataModelFromVersioned,
+				ResponseConverter: converter.RadiusPlaneDataModelToVersioned,
+			})
+	}
```
Do we need a default case here?

rynowak replied:
I will add a default case that returns an error.

ytimocin commented on pkg/ucp/integrationtests/planes/aws_test.go:
```diff
@@ -73,7 +73,6 @@ func Test_AWSPlane_GET_Found(t *testing.T) {
 }
 
 func Test_AWSPlane_LIST(t *testing.T) {
-	t.Skip("This functionality is currently broken. See https://github.com/radius-project/radius/issues/4878")
```
🧐

ytimocin commented on swagger/specification/ucp/resource-manager/UCP/preview/2023-10-01-preview/examples/AwsPlanes_List.json:
```diff
@@ -0,0 +1,24 @@
+  "responses": {
+    "200": {
+      "body": {
+        "value": [
```
Was it `values`?

rynowak replied:
If it was, that would be wrong.

ytimocin replied:
https://github.com/radius-project/radius/pull/7499/files#r1591556066

ytimocin commented on swagger/specification/ucp/resource-manager/UCP/preview/2023-10-01-preview/examples/AzureCredentials_CreateOrUpdate.json:
```diff
@@ -0,0 +1,57 @@
+{
+  "operationId": "AzureCredentials_CreateOrUpdate",
+  "title": "Create or update a Azure credential",
```
nit: 
```suggestion
  "title": "Create or update an Azure credential",
```

ytimocin commented on swagger/specification/ucp/resource-manager/UCP/preview/2023-10-01-preview/examples/AzureCredentials_List.json:
```diff
@@ -0,0 +1,32 @@
+  "responses": {
+    "200": {
+      "body": {
+        "value": [
```
Same question here (`values`)?

rynowak replied:
It should be `value`: https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#collections

All of our list APIs use `value` as the property name. If they don't then it's a bug.

ytimocin replied:
https://github.com/radius-project/radius/pull/7499/files#r1591556066

rynowak commented on deploy/Chart/crds/radius/radapp.io_recipes.yaml:
```diff
@@ -3,7 +3,8 @@ apiVersion: apiextensions.k8s.io/v1
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.15.0
+    controller-gen.kubebuilder.io/version: v0.9.1
```
oopsie-doopsie.

rynowak commented on pkg/ucp/api/v20231001preview/common.go:
```diff
@@ -0,0 +1,42 @@
+)
+
+func fromProvisioningStateDataModel(state v1.ProvisioningState) *ProvisioningState {
+	converted := ProvisioningState(string(v1.ProvisioningStateSucceeded))
```
Yes. The plane APIs don't currently do any async processing, so this will always have the default right now.

rynowak commented on pkg/ucp/api/v20231001preview/radiusplane_conversion.go:
```diff
@@ -0,0 +1,70 @@
+				Location: to.String(src.Location),
+				Tags:     to.StringMap(src.Tags),
+			},
+			InternalMetadata: v1.InternalMetadata{
```
Provisioning state is **readonly**. It must not be copied from the versioned model -> internal model.

rynowak committed:
Redesign plane resource types

This change splits the plane resource type into multiple types for maintainability as reviewed in radius-project/design-notes#42. See the design document for the motivation and dicussion of the specific design details. In summary, previously we had one resource type for all planes, and now we have a few separate ones.

This is a technically a breaking change but the impact on users is very low because it's mainly UCP itself that interacts with the planes.

The majority of the code changes are due to:

    Simplifying the resource group client API
    Updating and adding examples

Signed-off-by: Ryan Nowak <nowakra@gmail.com>

rynowak commented:
@ytimocin - updated

ytimocin commented on pkg/ucp/integrationtests/planes/testdata/awsplane_v20231001preview_list_responsebody.json:
```diff
@@ -1,12 +1,14 @@
 {
-    "values": [
```
I knew I saw **values** somewhere. You changed it. Thank you!

rynowak replied:
Examples are pretty frequently wrong, because they are manual 😢

rynowak commented:
Thanks @ytimocin !
